SU2_CFD/include/gradients/computeGradientsGreenGauss.hpp

```diff
@@ -0,0 +1,189 @@
+ * (http://su2foundation.org)
+ *
+ * Copyright 2012-2019, SU2 Contributors (cf. AUTHORS.md)
+ *
+ * SU2 is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * SU2 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with SU2. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "../../../Common/include/omp_structure.hpp"
+
+
+/*!
+ * \brief Compute the gradient of a field using the Green-Gauss theorem.
+ * \note Gradients can be computed only for a contiguous range of variables, defined
+ *       by [varBegin, varEnd[ (e.g. 0,1 computes the gradient of the 1st variable).
+ *       This can be used, for example, to compute only velocity gradients.
+ * \note The function uses an optional solver object to perform communications, if
+ *       none (nullptr) is provided the function does not fail (the objective of
+ *       this is to improve test-ability).
+ * \param[in] solver - Optional, solver associated with the field (used only for MPI).
+ * \param[in] kindMpiComm - Type of MPI communication required.
+ * \param[in] kindPeriodicComm - Type of periodic communication required.
+ * \param[in] geometry - Geometric grid properties.
+ * \param[in] config - Configuration of the problem, used to identify types of boundaries.
+ * \param[in] field - Generic object implementing operator (iPoint, iVar).
+ * \param[in] varBegin - Index of first variable for which to compute the gradient.
+ * \param[in] varEnd - Index of last variable for which to compute the gradient.
+ * \param[out] gradient - Generic object implementing operator (iPoint, iVar, iDim).
+ */
+template<class FieldType, class GradientType>
+void computeGradientsGreenGauss(CSolver* solver,
+                                MPI_QUANTITIES kindMpiComm,
+                                PERIODIC_QUANTITIES kindPeriodicComm,
+                                CGeometry& geometry,
+                                CConfig& config,
+                                const FieldType& field,
+                                size_t varBegin,
+                                size_t varEnd,
+                                GradientType& gradient)
+{
+  size_t nPointDomain = geometry.GetnPointDomain();
+  size_t nDim = geometry.GetnDim();
+
+#ifdef HAVE_OMP
+  constexpr size_t OMP_MAX_CHUNK = 512;
+
+  size_t chunkSize = computeStaticChunkSize(nPointDomain,
+                     omp_get_max_threads(), OMP_MAX_CHUNK);
+#endif
+
+  /*--- Start OpenMP parallel section. ---*/
+
+  SU2_OMP_PARALLEL
```

Comment:
This starts a parallel section, the code within {} that comes after is executed simultaneously by the threads.